### PR TITLE
Distance Sensor : Driver for Lanbao CM8JL65

### PR DESCRIPTION
**Pull request by cmic0**
Driver for a new distance sensor (Lanbao PSK-CM8JL65-CC5).

**Additional context**
If using TELEM1 port, disable MAVLINK messages there (MAV_0_CONFIG PARAMETER).
Data is published on distance_sensor topic at 10Hz rate. 

**Test data / coverage**
image

### CODE REVIEW

**Comment by cmic0** [comment]
@DanielePettenuzzo @RomanBapst FYI

**Comment by cmic0** [comment]
@dagar I get build failures on things not related to this driver. Could you check please?

**Comment by mhkabir** [comment]
This should be using the new serial port allocation interface.

File: ROMFS/px4fmu_common/init.d/rc.sensors [hunk 1]
```diff
@@ -93,6 +93,20 @@ then
 	sf1xx start -a
 fi
 
+# Lanbao PSK-CM8JL65-CC5 distance sensor
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.sensors]: Please use the generic UART configuration instead - see [this commit](https://github.com/PX4/Firmware/commit/b5e552924a998291820dd3741ddff41899dacbc5) for example (in particular the module.yaml).

**Comment by dagar** [comment]
> @dagar I get build failures on things not related to this driver. Could you check please?

The fixedwing mission test has somewhat recently started failing intermittently. I'll restart it for now.

File: ROMFS/px4fmu_common/init.d/rc.sensors (hunk 1 above)
Comment (cmic0) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: Upgraded driver startup with generic UART conf ([see here](https://github.com/Auterion/Firmware/commit/3650e6d88e4239aab3438ba3b1391024b76894ee)).
Tested on TELEM1 and TELEM2.

File: src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp [hunk 2]
```diff
@@ -0,0 +1,784 @@
...
+	if (collect_ret == -EAGAIN) {
+		_cycle_counter++;
+		/* We are missing bytes to complete the packet, re-cycle at 1ms */
+		//	work_queue(HPWORK,&_work,(worker_t)&CM8JL65::cycle_trampoline,this,USEC2TICK(1000LL));
+		//	return;
+	}
+
+
+	/* schedule a fresh cycle call when a complete packet has been received */
+	//work_queue(HPWORK,&_work,(worker_t)&CM8JL65::cycle_trampoline,this,USEC2TICK(_conversion_interval - _cycle_counter * 1000LL));
```
Comment (bkueng) [inline src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp]: Please remove commented code (above as well).

File: src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp [hunk 3]
```diff
@@ -0,0 +1,784 @@
...
+}
+
+void
+CM8JL65::cycle()
+{
+	//PX4_DEBUG("CM8JL65::cycle() - in the cycle");
+	/* fds initialized? */
+	if (_fd < 0) {
+		/* open fd */
+		_fd = ::open(_port, O_RDWR);
```
Comment (bkueng) [inline src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp]: This is not good. It means that the `read` call will block if there is no data available, and therefore will block the entire work queue.
Please change this to:
```
_fd = ::open(_port, O_RDWR | O_NOCTTY | O_NONBLOCK);
```

File: src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp [hunk 4]
```diff
@@ -0,0 +1,784 @@
...
+	float				             _max_distance;
+	int         	             _conversion_interval;
+	work_s				             _work{};
+	ringbuffer::RingBuffer	  *_reports;
+	int				               _fd;
+	uint8_t			             _linebuf[25];
+	uint8_t                   _cycle_counter;
+
+	enum CM8JL65_PARSE_STATE	 _parse_state;
+	unsigned char             _frame_data[4];
```
Comment (bkueng) [inline src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp]: Can you use a define for the length here and use it for the parsing as well? `int cm8jl65_parser(uint8_t c, uint8_t parserbuf[PARSER_BUF_LENGTH], CM8JL65_PARSE_STATE *state, uint16_t *crc16, int *dist)`

File: src/drivers/distance_sensor/cm8jl65/cm8jl65_parser.cpp [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+	while (crc16_length--) {
+		i = crc_low_byte ^ *(dataFrame++);
+		crc_low_byte = (unsigned char)(crc_high_byte ^ crc_msb_vector[i]);
+		crc_high_byte = crc_lsb_vector[i];
+	}
+
+	return (unsigned short)(crc_high_byte << 8 | crc_low_byte);
+}
+
+int cm8jl65_parser(uint8_t c, uint8_t *parserbuf, CM8JL65_PARSE_STATE *state, uint16_t *crc16, int *dist)
```
[the remark this answers is not in the document]
Comment (cmic0) [reply src/drivers/distance_sensor/cm8jl65/cm8jl65_parser.cpp]: Added passing by reference even for dist variable.

File: src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp (hunk 3 above)
Comment (cmic0) [reply src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp]: Good to know, thanks!

**Comment by cmic0** [comment]
@bkueng thanks for the review! Applied your suggestions, now should be ok.. let me know :-)!

**Comment by dagar** [comment]
You'd need to rebase to get the Jenkins hardware test to pass, but it isn't relevant for this PR.

**Comment by dagar** [comment]
Is this a standard CRC? I can't seem to find the datasheet for this device.
Consider the existing CRCs in systemlib, leddar_one, rc sumd, micrortps, or even NuttX itself.

**Comment by cmic0** [comment]
@dagar rebased on upstream, CI passed!
About the CRC: I got the infos from the manufacturer. It is a CRC16 IBM 8005. The two lookup tables defined in the code are standard 8005 ones, just splitted into MSB and LSB bytes. CRC calculation via lookup tables should be faster at the expense of some more memory occupation (not a CRC16 guru.. google helped me :-) ).

**Pull request merged**